common/TTbarReconstruction/recommendedSmearing/SmearingFunction.h

```diff
@@ -0,0 +1,343 @@
+#pragma once
+
+#define _USE_MATH_DEFINES
+#include <cmath>
+#include <string>
+#include <random>
+#include <functional>
+
+#include <Math/VectorUtil.h>
+#include <Math/Vector4D.h>
+#include <Math/Vector3Dfwd.h>
+#include <Math/AxisAngle.h>
+#include <TH2.h>
+#include <TH1.h>
+#include <TFile.h>
+
+/* 
+ * Virtual base class for all smearing functions.
+ * Has two calls, and two modes:
+ *  -- Modes:
+ *      - Unweighted: the solutions returned follow the PDF, the weight returned is 1. Warning: if the efficiency is very low, no check is made that the number of attempts in the unweighting be reasonable!
+ *      - Weighted: the solutions returned are uniformly distributed, the weight returned is the PDF evaluated on the solution
+ *  -- Calls:
+ *      - Vector: an input vector is given, and transformed to give the output vector
+ *      - Scalar: either an output scalar is built, or the TF is simply evaluated on an input
+ *
+ */
+class SmearingFunction {
+  public:
+    
+    SmearingFunction(): myPSgen(0, 1) {}
+    virtual ~SmearingFunction() {}
+
+    virtual double Evaluate(const myLorentzVector& input, myLorentzVector& output, const bool unweighted = true) { return 0;}
+    virtual double Evaluate(double& output, const bool unweighted) { return 0; } // No default value here since it would be ambiguous with this second overload:
+    virtual double Evaluate(const double input) { return 0; }
+      
+    double getRandom() { return myPSgen(engine); }
+
+  private:
+
+    std::mt19937_64 engine;
+    std::uniform_real_distribution<double> myPSgen;
+};
+
+
+class DiracDelta: public SmearingFunction {
+
+  public:
+
+    DiracDelta(): value(0) {}
+    DiracDelta(double value): value(value) {}
+    virtual ~DiracDelta() {}
+
+    virtual double Evaluate(const myLorentzVector& input, myLorentzVector& output, const bool unweighted = true) {
+      output = input;
+      return 1;
+    }
+    
+    virtual double Evaluate(double& output, const bool unweighted) {
+      output = value;
+      return 1;
+    }
+
+    virtual double Evaluate(const double input) {
+      return input == value;
```

Comment:
Testing equality of two double like this is error prone. You should do something like `std::abs(input - value) < PRECISION`